### ISSUE TITLE
Generating Type Definitions from Javascript is blocked by requiring private name

### ISSUE THREAD

**Issue by nbbeeken** [microsoft/TypeScript#37832]
**TypeScript Version:**  3.9.0-dev.20200407

**Search Terms:**
- TS9006
- Declaration emit for this file requires using private name
- Explicit type annotation 
- Unblock declaration emit.

**Code**
```js
// thing.js // maybe something from deep in my library
'use strict';
class Thing {}
module.exports = { Thing };
```
```js
// index.js
"use strict";
const Thing = require("./thing").Thing;
module.exports = { Thing }; // re-export to the world
```
tsconfig.json:
```json
{
  "compilerOptions": {
    "allowJs": true,
    "checkJs": true,
    "declaration": true,
    "emitDeclarationOnly": true,
    "declarationDir": "types",
    "strict": false,
    "target": "ES2018",
    "module": "commonJS",
    "moduleResolution": "node",
    "lib": ["ES2018"]
  },
  "include": ["index.js", "thing.js"]
}
```

**Expected behavior:**
I believe I should get two declaration files out of this that would look something like:
```ts
// thing.d.ts
export class Thing {}
```
```ts
// index.d.ts
import { Thing } from './thing';
export { Thing };
```
**Actual behavior:**
`index.d.ts` fails to be emitted with the error:
```
TS9006: Declaration emit for this file requires using private name 'Thing' from module '"/Users/neal/Desktop/privateNameTSC/thing"'. An explicit type annotation may unblock declaration emit.
```
It's not clear to me how to annotate the module.exports to make the "name" not private anymore
Thanks for taking a look.

**Playground Link:** Can't because imports related but I made a minimum repro here: https://github.com/nbbeeken/private-name-error-tsc

**Related Issues:** #9865

**Comment by sandersn** [comment]
This is broken in 3.8 and 3.9 at least.

**Comment by mbroadst** [comment]
@sandersn thank you kindly for the quick triage. This is a bit of a roadblock for our migration to TypeScript, would it be helpful if we tried to fix this locally? Or is there already someone on your end that will begin work shortly?

**Comment by sandersn** [comment]
From talking to @weswigham about a similar issue (probably the same issue), it's an unfortunate limitation in the way that we bind `module.exports = { Thing }`; although it looks like it could be translated to `export { Thing }`, we actually have to treat it as

```ts
declare const _tmp: { Thing: Thing }
export = _tmp
```

It's probably possible to special-case this exact pattern and emit `export { Thing }`.
@weswigham if you can weigh in, you know more about this than I do.

@mbroadst I put this in 4.0 assuming that the real fix would be complex.

**Comment by weswigham** [comment]
@sandersn you have it exactly.

**Comment by jamesopti** [comment]
Is there a workaround for this to prevent the CLI from erroring?

I tried `"checkJs": false,` but to no avail

**Comment by Anton-Loginov** [comment]
@nbbeeken Try to `"allowJs": false`, I think it helps you.

**Comment by nbbeeken** [comment]
> @nbbeeken Try to `"allowJs": false`, I think it helps you.

I appreciate the help! unfortunately that doesn't seem to work. The goal is to generate types from jsDoc comments so tsc needs to allow .js files as input. Here's the errors you get if allowJs is false: (setting checkJs to false just removes the TS5052 error)

```
error TS5052: Option 'checkJs' cannot be specified without specifying option 'allowJs'.

error TS18003: No inputs were found in config file './types/tsconfig.json'. Specified 'include' paths were '["index.js","thing.js"]' and 'exclude' paths were '["types"]'.
```

**Comment by achimmihca** [comment]
I am having the same error message without using explicit exports (tested with TypeScript 3.9.7). 
We have different implementations of classes with same name in different folders (for different customers). Although this results in a naming conflict, we would be able to @ts-ignore the duplicate declarations.

Sadly, some files produce an error that cannot be ignored: `Error:(1, 1) TS9005: Declaration emit for this file requires using private name 'Bla'. An explicit type annotation may unblock declaration emit.`

Minimal example:
```
// test1/Bla.js
class Bla extends BlaBase
{

}
```
```
// test2/Bla.js
class Bla extends BlaBase
{

}
```
```
// test3/BlaBase.js
class BlaBase
{
    constructor(name)
    {
        this.name = name;
    }
}
```

The error message is somewhat brittle. In this toy example it goes away when adding constructors to the two Bla implementations.
However, in our real code I am unsure what is causing the issue (e.g. a specific method or field).
Could you explain what is causing the message here? Any ideas to work around the issue?

**Comment by jrnail23** [comment]
@nbbeeken, I had a case somewhat similar to yours, where our app was exporting a function that set a property (which I think TypeScript recognizes as a constructor):
```js
module.exports = function(key) {
    const myImpl = implementations[key];
    this.invoke = myImpl.invoke;
};
```
and it was throwing this error:
`error TS9005: Declaration emit for this file requires using private name 'exports'. An explicit type annotation may unblock declaration emit.`

By changing `module.exports` to `exports`, the compiler error went away.

**Comment by nbbeeken** [comment]
Thanks for the suggestion @jrnail23!
I gave that a go but now there are type errors about `exports` not existing:
```
index.js:3:23 - error TS2306: File '.../test-ts-bug/thing.js' is not a module.
3 const Thing = require("./thing").Thing;
                        ~~~~~~~~~
thing.js:4:1 - error TS2304: Cannot find name 'exports'.
4 exports = { Thing };
  ~~~~~~~
Found 2 errors.
```
Here's a quick screenshot to show I'm using the example in the issue description.
<details>
<summary>Screenshot:</summary>
<br>
<img src="https://user-images.githubusercontent.com/10873993/93492346-b1554000-f8d8-11ea-99b8-1eda6c779ebe.png" />
</details>

**Comment by molisani** [comment]
Came to this issue because I was getting the same error, even when I had `declaration: false`. I just tested the repro (https://github.com/nbbeeken/private-name-error-tsc) with `typescript@beta` (`4.1.0-beta`) and it no longer produces this error.

**Comment by sandersn** [comment]
Yep, this is fixed in 4.1 -- although I found a related bug when `require`ing a relative path like `"./thing"` instead of `"fs"`.

**Comment by ggreco** [comment]
Are you sure this is fixed in 4.1?

`npm install typescript@beta`

I have the same problem both on 4.0.3 and 4.1, not sure is the exact same problem, minimal example included here:

https://stackoverflow.com/questions/64460213/create-typescript-declaration-files-for-existing-jsdoc-annotated-commonjs-librar

... maybe I did something wrong? Should I open a new ticket with the code I posted in stackoverflow?

**Comment by Bnaya** [comment]
re-exporting classes now leads to tsc crash "Unhandled alias declaration kind in symbol serializer!"
I've filed a bug report:
https://github.com/microsoft/TypeScript/issues/41182

**Comment by glen-84** [comment]
@sandersn @weswigham – Can this issue be reopened?

I get the following when trying to generate definitions for stylelint (using TS `4.1.0-dev.20201025`):

```
lib/assignDisabledRanges.js:1:1 - error TS9006: Declaration emit for this file requires using private name 'CommentExt' from module '"postcss/lib/comment"'. An explicit type annotation may unblock declaration emit.

1 'use strict';
  ~~~~~~~~~~~~

lib/descriptionlessDisables.js:1:1 - error TS9006: Declaration emit for this file requires using private name 'CommentExt' from module '"postcss/lib/comment"'. An explicit type annotation may unblock declaration emit.

1 'use strict';
  ~~~~~~~~~~~~

lib/needlessDisables.js:1:1 - error TS9006: Declaration emit for this file requires using private name 'CommentExt' from module '"postcss/lib/comment"'. An explicit type annotation may unblock declaration emit.

1 'use strict';
  ~~~~~~~~~~~~

lib/utils/parseCalcExpression/parser.js:383:13 - error TS9005: Declaration emit for this file requires using private name 'Parser'. An explicit type annotation may unblock declaration emit.

383             var parser = (function () {
                ~~~
```

**Comment by Bnaya** [comment]
@glen-84 can you please share a minimal repro?

**Comment by glen-84** [comment]
@Bnaya,

You can just clone the [stylelint](https://github.com/stylelint/stylelint) repo, update TS, and add the following to the `tsconfig.json` file:

```json
        "declaration": true,
	"emitDeclarationOnly": true,
	"outDir": "types-new"
```

(and optionally remove `"noEmit": true`, which is not needed)

Then run `npx tsc`.

**Comment by Bnaya** [comment]
@glen-84 While the error is the same, it's not the same issue,
see my workarounds there:
https://github.com/stylelint/stylelint/pull/5004/files
(types are now emitted)

What we are missing there to make it nicer is a way to `export type` from ts-jsdocs

**Comment by Bnaya** [comment]
> So this is triggered by the use of @typedef to define a type, instead of importing the type each time? Should it work?

I believe something like that

> Do you know if there is an existing GitHub issue for this? 

I think not

> Where do you see that?

Example: `import('stylelint').DisabledRange`

**Comment by glen-84** [comment]
> I think not

Okay, I'll probably open one.

> Example: import('stylelint').DisabledRange

This resolves to the local types in the `types` directory, not an `@types` package. These types would likely be removed once they can be generated automatically via TypeScript.

**Comment by Bnaya** [comment]
> This resolves to the local types in the types directory, not an @types package. These types would likely be removed once they can be generated automatically via TypeScript.

I'de say to relocate them, and import them as regular source files, and not like so (It will break if you publish)
But let's stop hijacking this issue, if there's somewhere else you are working on this, i'de love to help

**Comment by ggreco** [comment]
Here is a minimal example, less than 20 lines in total, it gives the error of this bug both on typescript 4.0 that on 4.1 beta:

https://github.com/ggreco/doc2dec

- checkout the repo
- npm install
- npm run build

**Comment by glen-84** [comment]
@ggreco Will you create a new issue?

**Comment by Bnaya** [comment]
> Here is a minimal example, less than 20 lines in total, it gives the error of this bug both on typescript 4.0 that on 4.1 beta:
> 
> https://github.com/ggreco/doc2dec
> 
> * checkout the repo
> * npm install
> * npm run build

Ugly workaround:
`use import('./rectangle').Rectangle`

**Comment by ggreco** [comment]
> @ggreco Will you create a new issue?

I created a new bug for this, https://github.com/microsoft/TypeScript/issues/41250

@Bnaya thanks for the workaround, it works, but it totally destroy the jsdoc markup for javascript users of the library...

**Comment by weswigham** [comment]
> Additionally, this bug still persist, see this repro:
> https://gist.github.com/Bnaya/f88130a9321b2706e1745f8695ff9664
> 
> When trying to avoid the above crash, i've set the imported class to an intermediate variable, and exported it.
> that leads again to the "using private name" error

This is fixed in latest `master` (likely by a combination of the other fixes going in) and now emits
```ts
//// [lib.d.ts]
/**
 * @param {string} a
 */
export function bar(a: string): string;
export class SomeClass {
    a(): number;
}
//// [main.d.ts]
export const IntermediateClass: typeof SomeClass;
import { SomeClass } from "./lib";
```

If anyone comes back to this thread using latest TS and thinks they have "the same bug" - please open a new thread with your repro - tracking stuff in closed issues is hard - thanks~

**Comment by tttinkl** [comment]
I  used `require(./xxx.js)` to skip this error; but i don't konw why it works.

**Comment by rajsite** [comment]
> @nbbeeken, I had a case somewhat similar to yours, where our app was exporting a function that set a property (which I think TypeScript recognizes as a constructor):
> 
> ```js
> module.exports = function(key) {
>     const myImpl = implementations[key];
>     this.invoke = myImpl.invoke;
> };
> ```
> 
> and it was throwing this error: `error TS9005: Declaration emit for this file requires using private name 'exports'. An explicit type annotation may unblock declaration emit.`
> 
> By changing `module.exports` to `exports`, the compiler error went away. I'm not sure if this actually worked, but it did get rid of the error. 🤷‍♂️

I am seeing a similar error message in TypeScript 4.3.2 with the following pattern:

```ts
/* data_listener.js */
export function createDataListener() {
    return async function dataListener(someObj, x, y) {
        // `this` references the model instance
       this.blah = something;
    };
}

/* create.js */
import {createDataListener} from "../data_listener";
someObj.setDataListener(createDataListener().bind(model, someObj));
```

I tried removing the `this` reference as  a hint from the quoted comment (maybe TypeScript thinks the returned function is being used as a constructor function when it is not) and I'm able to compile that code:

```ts
/* data_listener.js */
export function createDataListener() {
    return async function dataListener(model, someObj, x, y) {
       model.blah = something;
    };
}

/* create.js */
import {createDataListener} from "../data_listener";
someObj.setDataListener(createDataListener().bind(undefined, model, someObj));
```
